# Crash course in Python basics

## Execute these cells but otherwise ignore this section for now

In [ ]:
from __future__ import print_function
from __future__ import division

# Interacting with Python in Jupyter

In [ ]:
print(1234)

1234


Comment on the parentheses:
- `print` is the command, and within the parentheses we specify what is to be printed
- this is a general theme, there are commands (functions), and they have arguments, which are specified in parentheses

Cell execution in Jupyter/Colab: You can edit cells and re-execute. Try it: edit the cell below to change the numbers in the parentheses, and re-execute.

Note: Jupyter and Colab are not the same thing, but we will treat the terms interchangably for the time being

In [ ]:
print(2 + 98234785)

98234787


Note that we can already do math! We are adding numbers. We can also do:

- Subtraction: `-`
- Multiplication: `*`
- Division: `/`

In [ ]:
print(2-7)

-5


In [ ]:
print(10*26)

260


In [ ]:
print(6/5)

1.2


Can put multiple commands in one "cell" in Jupyter/Colab.

Let's do two additions in one cell:

In [ ]:
print(2 + 2)
print(100 + 100)

4
200



You can also get results without `print` calls---e.g. just typing `2+2` will print out the result:

In [ ]:
2+2

4

But this only works for the last command in a group of commands. E.g. here, we don't see the result of `2+2` but we see the result of `10+10`:

In [ ]:
2+2
10+10

20

It is better practice to use `print` always.

We can also add human-readable comments in code cells by using the `#` sign.   Anything that follows the `#` in a line is ignored by Python.

In [ ]:
# here is a note to myself: this command will compute an approximation to pi
print(22/7)

3.142857142857143


Jupyter notebook cells vs "base Python":
- Pure Python code is like one big code cell of Jupyter, and nothing else
- The notebook interface allows us to insert nicely formatted text (such as this), or images, etc. in between lines of Python code

# Packages, imports, methods

- Python (i.e. base Python language)doesn't have a command for taking square roots
- But some nice people implemented square roots (and other things) for us, and bundled them inside a "**package**"
- Our package is called `numpy`: This is one of the core packages we will use in this course

- We load (or activate, in our current notebook session) a package using the `import` command
- Within a given package, we choose the needed command using a dot (see below)

In [ ]:
import numpy

In [ ]:
numpy.sqrt(4)

2.0

The square root command (function) is called `sqrt`:

In [ ]:
print(numpy.sqrt(4))

2.0


In [ ]:
print(numpy.sqrt(2))

1.4142135623730951


- It is possible to import a package by specifying a different name than its original name
- The syntax is "`import <original package name> as <preferred name>`"
- This is done for e.g. brevity

NumPy is commonly imported as `np`:

In [ ]:
import numpy as np

From now on, we refer to `numpy` as `np`:

In [ ]:
print(np.sqrt(9))

3.0


In [ ]:
np.sqrt(np.sqrt(4))

1.4142135623730951

Numpy has lots of other functionality which we will explore over time.

To have a quick teaser, we list the many commands available under `numpy` using the `dir` command:

In [ ]:
# the dir function lists all the available commands (and other things) under a package
dir(np)

['ALLOW_THREADS',
 'AxisError',
 'BUFSIZE',
 'CLIP',
 'ComplexWarning',
 'DataSource',
 'ERR_CALL',
 'ERR_DEFAULT',
 'ERR_IGNORE',
 'ERR_LOG',
 'ERR_PRINT',
 'ERR_RAISE',
 'ERR_WARN',
 'FLOATING_POINT_SUPPORT',
 'FPE_DIVIDEBYZERO',
 'FPE_INVALID',
 'FPE_OVERFLOW',
 'FPE_UNDERFLOW',
 'False_',
 'Inf',
 'Infinity',
 'MAXDIMS',
 'MAY_SHARE_BOUNDS',
 'MAY_SHARE_EXACT',
 'ModuleDeprecationWarning',
 'NAN',
 'NINF',
 'NZERO',
 'NaN',
 'PINF',
 'PZERO',
 'RAISE',
 'RankWarning',
 'SHIFT_DIVIDEBYZERO',
 'SHIFT_INVALID',
 'SHIFT_OVERFLOW',
 'SHIFT_UNDERFLOW',
 'ScalarType',
 'Tester',
 'TooHardError',
 'True_',
 'UFUNC_BUFSIZE_DEFAULT',
 'UFUNC_PYVALS_NAME',
 'VisibleDeprecationWarning',
 'WRAP',
 '_CopyMode',
 '_NoValue',
 '_UFUNC_API',
 '__NUMPY_SETUP__',
 '__all__',
 '__builtins__',
 '__cached__',
 '__config__',
 '__deprecated_attrs__',
 '__dir__',
 '__doc__',
 '__expired_functions__',
 '__file__',
 '__getattr__',
 '__git_version__',
 '__loader__',
 '__name__',
 '__package__',
 '__path__',
 

# Variables

Variables are like named boxes where we put data.

A simple example: We put the number `123` inside a box named `a`:

In [ ]:
a = 123

From now on, "`a`" now refers to the number `123`:

In [ ]:
print(a)

123


In [ ]:
print(a+1)

124


Another variable:


In [ ]:
b = 10000

In [ ]:
b

10000

In [ ]:
print(a+b)

10123


You can refer to variables while setting variables, too

In [ ]:
c = a + b

In [ ]:
print(c)

10123


- Note that `c` is set to the current value of the right hand side, i.e., if you update `a` or `b`, `c` will not change, it will stay as `133`.
- There are Python packages that allow you to link variables to each other so that if you change one, the others change, too, but let's leave that to the future.

Here's something that is perhaps confusing at first sight:

In [ ]:
a = 2000

In [ ]:
c

10123

In [ ]:
a = a + 1

- What does this even mean? It is mathematically inconsistent!
- The `=` sign is perhaps a bit inappropriate; in Python, it is used for:
   - setting the thing on the left hand side to (the current value of) whatever is on the right hand side
- In some other languages, one uses a less ambiguous symbol for assignment
  - One example is the symbol `<-`, which is used in the R programming language

In the case above, `a` was originally 123, so this command sets `a` to `123 + 1`:

In [ ]:
print(a)

2001


- The variable `a` has increased by one, from now on, it is `124`

This also shows that we can update variables in a given Python session, they are not set for good.

Let's do that a few times in a single cell:

In [ ]:
b = 10
print(b)
b = b + 1
print(b)
b = b + 1
print(b)

10
11
12


- Up to now, we used `a`, `b`, etc. for variable names, but we can use longer names
- Some things aren't accepted as variable names
- You are allowed to use letters, numbers, and underscores in variable names
- Variable names can't *start* with a number
- Avoid using name that Python already knows about; e.g. don't name your variable `print`

In [ ]:
myvariable = 1
my_other_variable = 2
my_variable123 = 3

print(myvariable)
print(my_other_variable)
print(my_variable123)

1
2
3


In [ ]:
print(myvariable + my_other_variable)

3


- It is a good idea to give concise but descriptive names to variables
  - Can use `height`, `speed`, `mass`, etc. for physics problems
- Integers/variables that count things are usually given names like i, j, k
- Real number are usually given names like x, y, z
- If you will use a compound name that consists of multiple words, the Python practice is to use lowercase letters, and to separate words with single underscores (as in `my_other_variable`)

# Strings

- What if we wanted to print the letter "`a`", or any other letter, or some words, even sentences?
- We put such things in quote signs


In [ ]:
print("Hello!")

Hello!


In [ ]:
s = "Hello again!"

In [ ]:
print(s)

Hello again!


Such variables are called **strings** (as opposed to numbers)

We can use single quotes instead of double:

In [ ]:
print('Hey!')

Hey!


Python also has a notion of "adding" strings: it simply means concatenating them:

In [ ]:
t = "!!!!"

In [ ]:
print(s + t)

Hello again!!!!!


In [ ]:
u = s + t

In [ ]:
print(u)

Hello again!!!!!


# Errors

- Programming languages are very strict and precise
- You can't just type anything you want and hope the computer understands, you have to adhere to a restrictive set of rules
- If you don't, the computer will stop and (hopefully) complain

Much of our time as Python programmers will be spent trying to understand the error messages in Python, so it is a good idea to make a habit of paying attention to them.

Here is a nonsensical "command", which annoys Python:

In [ ]:
print mrint oyle bisey

SyntaxError: ignored

- The last line is what we should focus on
- `SyntaxError` is the type of error Python noticed: by this, Python means "I don't even know what you are talking about"
  - I.e., the specified command can't be interpreted within the Python language

Here's another error:

In [ ]:
s = "Let's see if this works"
a = 10
print(s + a)

TypeError: ignored

Once again, the last line is what we should focus on (at least initially)
- `TypeError`: we're trying to do something that doesn't make sense with the data types we have
- Mentions 'str' (string) and 'int' (integer) objects
- Python is telling us it doesn't know how to add an integer to a string

Another error:

In [ ]:
print(5/0)

ZeroDivisionError: ignored

# Other data types: Lists

- We saw numbers and strings
  - Numbers actually come in more than one type, but let's leave that for now
- A **list** holds a list of multiple values:
- We use brackets and commas to separate the values we want to include in our list

In [ ]:
l = [10, 20, 30, 40] # note that this is the "letter l", not the number one (and this is a comment since it comes after the # sign)

In [ ]:
print(l)

[10, 20, 30, 40]


You can access individual elements of the list using indexing with brackets
- We count elements starting from **zero**
- **not** one

In [ ]:
l[2]

30

In [ ]:
print(l[0])

10


In [ ]:
print(l[1])

20


This is a big, common source of confusion. Keep this in mind. Always double, triple check your indices for off-by-one errors.

Let's lear a new function, one that is relevant for lists
- To learn the length of a list, we use `len()`

In [ ]:
len(l)

4

Could've printed this, too, with `print(len(l))`

Let us print the last element of the list. The number of elements is `len(l)`, but since we start counting at zero, we subtract one to obtain the last element (what happens if we don't? try if you want):

In [ ]:
print(l[len(l)-1])

40


In [ ]:
l[-1]

40

We can change individual elements of a list using the (confusing) equals sign

In [ ]:
l[1] = 1000

In [ ]:
print(l)

[10, 1000, 30, 40]


The second element is now changed.

Can access the last element in different ways:

In [ ]:
print(l[3])

40


This one is maybe a bit of a surprise:

In [ ]:
print(l[-1])

40


In [ ]:
length_of_l = len(l)

In [ ]:
print(l[length_of_l - 1])

40


Now that we know two functions, print and len, let's say this again:
- Remember, don't use existing names as variable names

# Conditionals

- One of the core building blocks of programming, exists in all(?) languages
- Do something only if a condition holds
- And maybe do something else if the condition doesn't hold

Syntax (how it is done in the Python language):

In [ ]:
# let's use a variable
a = 10

In [ ]:
if a > 10:
    print("Your number is bigger than 10")
else:
    print("Your number is not bigger than 10")

Your number is bigger than 10


- **Indentation (space) is important in Python!**
- We first write "`if`", and follow with the logical condition we want to check
  - In this case, `a > 10`
  - No need for parentheses
- We then put a colon `:`
- On the next line, we put some space, and follow with the command we want executed if our condition holds
- Then put an `else:`, **without preceding space**
- And on the next line, put the command you want, once again after some space

Notes:

- We can use tabs instead of spaces
- Need to be consistent in our choice, though; if you put 4 spaces, then you have to continue using 4 spaces; can't switch between different numbers of spaces/tabs
- You can execute multiple commands if your condition holds by writing them one after the other, e.g.:

In [ ]:
a=10

In [ ]:
if a > 10:
    print("Your number is bigger than 10")
    aprint("isn't that funny?")
else:
    print("Your number is not bigger than 10")
    print("and it's not funny")

NameError: ignored

For the condition in an `if` statement, you can use various things that can be interpreted as "logical" values by Python

Let's see what "a>10" is really like:

In [ ]:
print(a>10)

False


This is not a string, i.e., it prints as "`False`", but it is not the string "False", it is a logical value.

In [ ]:
x = a < 100

In [ ]:
print(x)

True


You can set variables to logical values directly, you don't have to test for a specific condition

In [ ]:
x = True
# note that we did not use any quotes! this is not the string "True", it is the logical value True

In [ ]:
print(x)

True


Just as we can add or subtract numbers, we can perform logical operations on logical variables:

False

In [ ]:
print (x or (not x))

True


- We tested for an inequality. How about an equality?
- Can we say "`if x = 10`"?
- No, we saw that "=" has a different meaning; it is used for assignment
- For comparisons/equality tests, we use a pair of equals signs `==` (no space between)
  - A common mistake: using `=` when you intend to use `==`

In [ ]:
a = 10

if a == 10:
    print("your variable is ten")

your variable is ten


In [ ]:
a == 10

True

- We can nest if statements (an if inside an if)
- To do so, we need to keep track of indendation levels


Let's fill in the messages here

In [ ]:
x = 10
if x < 10:
    if x > 5:
        print("x is greater than 5 but less than 10")
    else:
        print("")
else:
    if x < 20:
        print("")
    else:
        print("")

Nesting like this is **not** recommended--makes code hard to read!

The "**elif**": Helps reduce the number of levels of indentation. Most easily shown with an example:

In [ ]:
x = 6

if x > 10:
  print("x is greater than 10")
elif x > 5:
  print("x is greater than 5, and less than or equal to 10")
elif x > 0:
  print("x is greater than 0, and less than or equal to 5")
else:
  print("x is less than or equal to 0")

x is greater than 5, and less than or equal to 10


# Exercise for the adventurous

Python has an `id` command that returns a unique identifier for each object. Let's try a few things:

In [ ]:
a=1

In [ ]:
b=2

In [ ]:
id(a)

140032584270064

In [ ]:
id(b)

140032584270096

The ids are different; these are different variables, each having its own id. Now look at this:

In [ ]:
c=1

In [ ]:
id(c)

140032584270064

In [ ]:
id(a) == id(c)

True

In [ ]:
a is c

True

Strange! These were supposed to be different variables, but they get the same id. Could it be that variables with the same value get the same id? Now look at this:

In [ ]:
x = 100000
y = 100000

In [ ]:
id(x)

140032156253136

In [ ]:
id(y)

140032156253200

In [ ]:
x is y

False

Even stranger! These guys actually get different ids.

Your task:
- Figure out which numbers result in two separate variables getting the same id
- What this means; i.e., what is happening under the hood

Googling is allowed!

In [ ]:
# integer values between -5 and 256 have its own unique address on the memory
# so python does not allocate a new address for the storage
x = 256
y = 256
print(id(x) == id(y))

True
